contracts/upgradeable_contracts/erc20_to_native/InterestConnector.sol

```diff
@@ -0,0 +1,286 @@
+    }
+
+    /**
+     * @dev Internal function for withdrawing some amount of the invested tokens.
+     * Reverts if given amount cannot be withdrawn.
+     * @param _token address of the token contract withdrawn.
+     * @param _amount amount of requested tokens to be withdrawn.
+     */
+    function _withdraw(address _token, uint256 _amount) internal {
+        if (_amount == 0) return;
+
+        uint256 invested = investedAmount(_token);
+        uint256 balance = _selfBalance(_token);
+        uint256 withdrawal = _amount > invested ? invested : _amount;
+
+        _withdrawTokens(_token, withdrawal);
+
+        uint256 redeemed = _selfBalance(_token) - balance;
+
+        // Make sure that at least withdrawal amount was indeed withdrawn
+        require(redeemed >= withdrawal);
+
+        _setInvestedAmount(_token, invested > redeemed ? invested - redeemed : 0);
+    }
+
+    /**
+     * @dev Internal function for setting minimum amount of tokens that cannot be invested.
+     * @param _token address of the token contract.
+     * @param _minCashThreshold minimum amount of underlying tokens that are not invested.
+     */
+    function _setMinCashThreshold(address _token, uint256 _minCashThreshold) internal {
+        uintStorage[keccak256(abi.encodePacked("minCashThreshold", _token))] = _minCashThreshold;
+    }
+
+    /**
+     * @dev Internal function for setting limits for paid interest amount in a single call.
+     * @param _token address of the token contract.
+     * @param _minInterestPaid minimum amount of interest paid in a single call.
+     * @param _maxInterestPaid maximum amount of interest paid in a single call.
+     */
+    function _setPaidInterestLimits(address _token, uint256 _minInterestPaid, uint256 _maxInterestPaid) internal {
+        require(_minInterestPaid < _maxInterestPaid);
+        uintStorage[keccak256(abi.encodePacked("minInterestPaid", _token))] = _minInterestPaid;
+        uintStorage[keccak256(abi.encodePacked("maxInterestPaid", _token))] = _maxInterestPaid;
+    }
+
+    /**
+     * @dev Internal function for setting interest receiver address.
+     * @param _token address of the invested token contract.
+     * @param _receiver address of the interest receiver.
+     */
+    function _setInterestReceiver(address _token, address _receiver) internal {
+        require(_receiver != address(this));
+        addressStorage[keccak256(abi.encodePacked("interestReceiver", _token))] = _receiver;
+    }
+
+    /**
+     * @dev Internal function for clamping the available interest amount.
+     * @param _token address of the invested token contract.
+     * @param _amount amount of underlying tokens available to withdrawal.
+     * @return amount in minInterestPaid...maxInterestPaid range.
+     */
+    function _clampInterest(address _token, uint256 _amount) internal view returns (uint256) {
+        require(_amount >= minInterestPaid(_token));
+        uint256 maxInterest = maxInterestPaid(_token);
```

Comment:
1) To limit possible losses in case of some attack
2) To limit the amount of funds that can be burnt/swapped at bad exchange rates (e.g. fee percentage for swap at Uniswap depends on the swapped amount)

Do you think that it won't be useful?

Reply:
> To limit possible losses in case of some attack

Do you have an example of such attack?

> To limit the amount of funds that can be burnt/swapped at bad exchange rates (e.g. fee percentage for swap at Uniswap depends on the swapped amount)

From my understanding this is not fees but a mechanism which protect the pool from drying out and depends on the size of the pool. So, in fact, this limit should be dynamic. Does it make sense to have the exchange rate difference limit instead of the interest limit? For example, if the exchange rate changes more than 2% for some particular interest amount, the transaction will revert.

Reply:
Ok, the exchange rate slippage seems like a better option. I have implemented it in the `InterestReceiverStakeBuyback` contract.

The only thing I am worried about, is that the call to `UniswapRouterV2.getAmountsOut` required for such check will add extra ~35k gas to the transaction cost.

Reply:
> I have implemented it in the InterestReceiverStakeBuyback contract.

Does it make sense to add this logic to `InterestReceiverSwapToETH` as well?

Reply:
I don't think so, WETH/DAI pair on Uniswap contains one of the largest pools. Swapping 100.000 DAI to ETH gives a price slippage of 0.15% only, which is extremely low.

Reply:
Things could change extremely fast